WordPress/Helpers/WPDBTrait.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * WordPress Coding Standard.
+ *
+ * @package WPCS\WordPressCodingStandards
+ * @link    https://github.com/WordPress/WordPress-Coding-Standards
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WordPressCS\WordPress\Helpers;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Util\Tokens;
+use PHPCSUtils\BackCompat\BCFile;
+use PHPCSUtils\Tokens\Collections;
+
+/**
+ * Helper utilities for sniffs which examine WPDB method calls.
+ *
+ * {@internal The method in this trait was previously contained in the
+ * `WordPressCS\WordPress\Sniff` class and has been moved here.}
+ *
+ * @package WPCS\WordPressCodingStandards
+ * @since   3.0.0
+ */
+trait WPDBTrait {
+
+	/**
+	 * Checks whether this is a call to a $wpdb method that we want to sniff.
+	 *
+	 * If available in the class using this trait, the $methodPtr, $i and $end properties
+	 * are automatically set to correspond to the start and end of the method call.
+	 * The $i property is also set if this is not a method call but rather the
+	 * use of a $wpdb property.
+	 *
+	 * @since 0.8.0
+	 * @since 0.9.0  The return value is now always boolean. The $end and $i member
+	 *               vars are automatically updated.
+	 * @since 0.14.0 Moved this method from the `PreparedSQL` sniff to the base WP sniff.
+	 * @since 3.0.0  - Moved from the Sniff class to this dedicated Trait.
+	 *               - The $phpcsFile parameter was added.
+	 *
+	 * {@internal This method should be refactored to not exhibit "magic" behaviour
+	 *            for properties in the sniff class(es) using it.}}
+	 *
+	 * @param \PHP_CodeSniffer\Files\File $phpcsFile      The file being scanned.
+	 * @param int                         $stackPtr       The index of the $wpdb variable.
+	 * @param array                       $target_methods Array of methods. Key(s) should be method name
+	 *                                                    in lowercase.
+	 *
+	 * @return bool Whether this is a $wpdb method call.
+	 */
+	protected function is_wpdb_method_call( File $phpcsFile, $stackPtr, $target_methods ) {
+
+		$tokens = $phpcsFile->getTokens();
+
+		// Check for wpdb.
+		if ( ( \T_VARIABLE === $tokens[ $stackPtr ]['code'] && '$wpdb' !== $tokens[ $stackPtr ]['content'] )
+			|| ( \T_STRING === $tokens[ $stackPtr ]['code'] && 'wpdb' !== strtolower( $tokens[ $stackPtr ]['content'] ) )
+		) {
+			return false;
+		}
+
+		// Check that this is a method call.
+		$is_object_call = $phpcsFile->findNext( Tokens::$emptyTokens, ( $stackPtr + 1 ), null, true );
+		if ( false === $is_object_call
+			|| isset( Collections::objectOperators()[ $tokens[ $is_object_call ]['code'] ] ) === false
+		) {
```

Comment:
Does the assigned variable name suggest that this could become a PHPCSUtils method - quite generic?
```php
$is_object_call = PHPCSUtils\Util\Foo::IsObjectCall( $stackPtr, $tokens ); // Or something similar.
```

Reply:
I mean, reading the rest of the method, there's nothing WPDB-special in it - that only happens in the early checks. I wonder if all of it could be turned into something like:

```
protected function is_wpdb_method_call( File $phpcsFile, $stackPtr, $target_methods ) {
	// Check if it's wpdb-related
	// Check if it's a method call via PHPCSUtils
}

Reply:
> Does the assigned variable name suggest that this could become a PHPCSUtils method - quite generic?

The assigned variable name is the same as it was in the original code in the `Sniff` class, so no, this is no indication of anything.

> I mean, reading the rest of the method, there's nothing WPDB-special in it - that only happens in the early checks

At this time, there is no intention for this to be moved to PHPCSUtils.

This method does weird and horrible stuff with properties in the (child) class using the method and there is no way I will ever allow that kind of code to get into PHPCSUtils.

There may be utilities in PHPCSUtils in the future which could simplify the code in this method, but I highly doubt it, as the code "walks tokens the wrong way" in my opinion (starts at the variable/class, not at the method call) and draws conclusions based on arbitrary local variable names (anything can be called `$wpdb`) which is unwarranted.

Aside from that, the property assignments being done require information which would be contained _within_ the method in PHPCSUtils and not returned, so that would mean that the method here would need to duplicate what PHPCSUtils would be doing, which doesn't make much sense from a performance point of view.

Reply:
Note: to be honest, as the sniffs using this method look at a specific parameter in a method call, I would much rather advocate for this method to be removed in the future in favour of a rewrite of those sniffs using an `AbstractFunctionCallParameter` base class combined with the PHPCSUtils `PassedParameters` class.

It's something I'd consider looking into when we switch to the abstracts in PHPCSUtils once available. If so, we can deprecate the method and remove it in the next major.